src/Akka.Hosting/ActorRegistry.cs

```diff
@@ -84,6 +134,72 @@ public bool TryGet(Type key, out IActorRef actor)
             return false;
         }
 
+        public async Task<IActorRef> GetAsync<TKey>(CancellationToken ct)
+        {
+            return await GetAsync(typeof(TKey), ct);
+        }
+
+        public async Task<IActorRef> GetAsync(Type key, CancellationToken ct)
+        {
+            // try to get the populated actor first, if available
+            if (TryGet(key, out var storedActor))
+            {
+                return storedActor;
+            }
+
+            var tcs = new TaskCompletionSource<IActorRef>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var waitingRegistration = new WaitForActorRegistration(key, tcs);
+            var registration = ct.Register(CancelWaiter(key, ct, waitingRegistration), _actorWaiters);
+            waitingRegistration.CancellationRegistration = registration;
+
+            _actorWaiters.AddOrUpdate(key,
+                type => ImmutableHashSet<WaitForActorRegistration>.Empty.Add(waitingRegistration),
+                (type, set) => set.Add(waitingRegistration));
+            
+
+            return await tcs.Task.ConfigureAwait(false);
+        }
+
+        private void NotifyWaiters(Type key, IActorRef value)
+        {
+            // remove the registrations and then iterate over them
+            if (_actorWaiters.TryRemove(key, out var registrations))
+            {
+                foreach (var r in registrations)
+                {
+                    r.Waiter.TrySetResult(value);
+                    r.CancellationRegistration.Dispose();
+                }
+            }
+        }
+
+        private static Action<object> CancelWaiter(Type key, CancellationToken ct, WaitForActorRegistration waitingRegistration)
+        {
+            return dict =>
+            {
+                // first step during timeout is to remove our registration
+                var d = (ConcurrentDictionary<Type, ImmutableHashSet<WaitForActorRegistration>>)dict;
```

Comment:
We use a blend of closures and passing in state objects via parameters here - I think this is probably fine.